src/pybind/mgr/dashboard/controllers/__init__.py

```diff
@@ -454,146 +541,107 @@ class RESTController(BaseController):
-        method = getattr(self, cherrypy.request.path_info.split('/')[-1])
-
-        if cherrypy.request.method not in ['GET', 'DELETE']:
-            method = RESTController._takes_json(method)
-
-        method = RESTController._returns_json(method)
-
-        cherrypy.response.status = 200
-
-        return method(*vpath, **params)
-
-    def _rest_request(self, is_element, *vpath, **params):
-        method_name, status_code = self._method_mapping[
-            (cherrypy.request.method, is_element)]
-        method = getattr(self, method_name, None)
-
-        if cherrypy.request.method not in ['GET', 'DELETE']:
-            method = RESTController._takes_json(method)
-
-        method = RESTController._returns_json(method)
-
-        cherrypy.response.status = status_code
-
-        return method(*vpath, **params)
-
-    @staticmethod
-    def _function_args(func):
-        return getargspec(func).args[1:]
-
-    @staticmethod
-    def _takes_json(func):
-        def inner(*args, **kwargs):
-            if cherrypy.request.headers.get('Content-Type',
-                                            '') == 'application/x-www-form-urlencoded':
-                return func(*args, **kwargs)
+    @classmethod
+    def endpoints(cls):
+        result = super(RESTController, cls).endpoints()
+        for _, val in inspect.getmembers(cls, predicate=callable):
+            if val.__name__ in cls._method_mapping:
+                meth = cls._method_mapping[val.__name__]
+
+                path = ""
+                if meth['resource']:
+                    res_id_params = cls.infer_resource_id()
+                    if res_id_params is None:
+                        raise TypeError("Invalid resource method '{}'. Please "
+                                        "check if method has declared "
+                                        "non-optional parameters to be used has"
+                                        " resource id, or specify the "
+                                        "RESOURCE_ID class property")
+                    res_id_params = ["{{{}}}".format(p) for p in res_id_params]
+                    path += "/{}".format("/".join(res_id_params))
+
+                val = cls._status_code_wrapper(val, meth['status'])
+                endp_func = cls.endpoint(meth['method'], path=path)(val)
+                result.append(cls.Endpoint(cls, endp_func))
+
+            elif hasattr(val, "_collection_method_"):
+                path = "/{}".format(val.__name__)
+                val = cls._status_code_wrapper(val, 200)
+                endp_func = cls.endpoint(val._collection_method_, path=path)(val)
+                result.append(cls.Endpoint(cls, endp_func))
+
+            elif hasattr(val, "_resource_method_"):
+                path = ""
+                res_id_params = cls.infer_resource_id()
+                if res_id_params is None:
+                    raise TypeError("Invalid resource method '{}'. Please "
```

Comment:
- The structure of the error message is a bit confusing.
- The string format parameter is missing, ``{}`` will be replaced by an empty string otherwise.